Platform/AlderlakeBoardPkg/Library/Stage1ABoardInitLib/Stage1ABoardInitLib.c

```diff
@@ -119,6 +121,86 @@ EarlyPlatformDataCheck (
   }
 }
 
+/**
+  CAR created by ACM covers only Initial code region.
+  This function will expand CAR to cover MRC data and improve performance of MRC code.
+
+**/
+VOID
+EFIAPI
+ExpandCacheAsRam (
+  VOID
+)
+{
+  MSR_IA32_MTRRCAP_REGISTER       MtrrCap;
+  MSR_IA32_MTRR_PHYSMASK_REGISTER MtrrMaskReg;
+  UINT64                          TempQword;
+  UINT64                          MemoryLength;
+  UINT64                          ValidMtrrAddressMask;
+  EFI_PHYSICAL_ADDRESS            MemoryAddress;
+  UINT32                          MtrrNumber;
+  UINT32                          VariableMtrrCount;
+  UINT8                           UsedMtrr;
+  UINT8                           Index;
+  UINT8                           PhysicalAddressBits;
+
+  // Get Valid MTRR Address Mask
+  PhysicalAddressBits  = GetPhysicalAddressBits ();
+  ValidMtrrAddressMask = (LShiftU64((UINT64) 1, PhysicalAddressBits) - 1) & (~(UINT64)0x0FFF);
+  // Current the number of Variable MTRRs
+  MtrrCap.Uint64 = AsmReadMsr64 (MSR_IA32_MTRRCAP);
+  VariableMtrrCount = MtrrCap.Bits.VCNT;
+  UsedMtrr = 0;
+  for (Index = 0; Index < VariableMtrrCount; Index++) {
+    MtrrNumber = MSR_CACHE_VARIABLE_MTRR_BASE + (Index << 1);
+    MemoryAddress = AsmReadMsr64 (MtrrNumber) & ValidMtrrAddressMask;
+    MtrrMaskReg.Uint64 = AsmReadMsr64 (MtrrNumber + 1);
+    if (MtrrMaskReg.Bits.V != 0) {
+      MemoryLength = ~(MtrrMaskReg.Uint64 & ValidMtrrAddressMask) + 1;
+      if (MemoryLength != 0) {
+        UsedMtrr++;
+      }
+    }
+  }
+
+  MtrrNumber = MSR_CACHE_VARIABLE_MTRR_BASE + (UsedMtrr << 1);
+
+  MemoryLength  = PcdGet32 (PcdMrcDataSize);
+  // Set MemoryLength to a valid MTRR length:
+  if (MemoryLength > SIZE_256KB) {
+    // Too large MRC data is unreasonable.
+    MemoryLength = SIZE_256KB;
+  } else if (MemoryLength < SIZE_4KB) {
+    // MTRR length must larger than 4KB.
+    MemoryLength = SIZE_4KB;
+  } else {
+    // Find the nearest Nth power of 2 that is not bigger than PcdMrcDataSize,
+    // for example, set length to 32KB if MRC data is 36KB.
+    for (Index = 0; Index < 18u; Index++) {
+      if ((MemoryLength & (SIZE_256KB >> Index)) != 0) {
+        MemoryLength = SIZE_256KB >> Index;
+      }
+    }
+  }
+  // Append MTRR to make sure MRCD cache-able, which can improve performance of FSP-M.
+  // Note: Mtrr base address MUST align to size (And size must be 2 to the Nth power and larger than 4KB),
+  // unfortunately SBL didnt promise this. So it must be mapped with multiple MTRRs.
+  MemoryAddress = (EFI_PHYSICAL_ADDRESS) ALIGN_DOWN (PcdGet32 (PcdMrcDataBase), MemoryLength);
+  while (MemoryAddress < (PcdGet32 (PcdMrcDataBase) + PcdGet32 (PcdMrcDataSize))) {
+    if (UsedMtrr >= VariableMtrrCount) {
+      break;
```

Comment:
There is a case here where it requires multiple MTRRs to cover the MRCD region, and the available MTRRs is at least one but insufficient to satisfy it. In this case, the code will exit with MTRRs only partially covering the MRC Data region. This may be acceptable, but I wanted to make you aware of it.

Reply:
Yes, I believe this is acceptable.
I tested this case, even if only part of the MRCD is covered, it will improve the performance of FSPM - the effect depends on the proportion of coverage, and there will be no impact on functionality.

Reply:
Yi,  i agree with Bejean,  We should avoid the case only map part of MRCD.  There's risk even result looks "positive" in your local.

Reply:
I think a possible risk is that my local test cannot cover all strange MRCD base and length pairs. Ok, I will update patch to check  valid MTRR number.